### PR TITLE
apiextensions-apiserver: stop cacher on CRD update

### LINKED ISSUE #53485: Leaking reflector metrics in 1.8
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Ran Kubernetes 1.8 apiserver for 3 days and the memory usage kept going up.

screenshot from 2017-10-05 10-12-19

Looking at a heap profile showed, that the majority of this memory is coming from the [`cache.NewNamedReflector`](https://github.com/kubernetes/client-go/blob/82aa063804cf055e16e8911250f888bc216e8b61/tools/cache/reflector.go#L107) function, or more specifically, the [metrics being instantiated in it](https://github.com/kubernetes/client-go/blob/82aa063804cf055e16e8911250f888bc216e8b61/tools/cache/reflector.go#L112).

[pprof.hyperkube.alloc_objects.alloc_space.inuse_objects.inuse_space.003.pb.gz](https://github.com/kubernetes/kubernetes/files/1358504/pprof.hyperkube.alloc_objects.alloc_space.inuse_objects.inuse_space.003.pb.gz)

I validated that this is indeed coming from metrics by checking the amount of metrics exposed by the `/metrics` endpoint.

```
$ curl localhost:8001/metrics | wc -l
49966
```

In an equivalent workload cluster running Kubernetes `v1.7.5`, this number is round `3000`.

So I took a look and noticed, that the majority of these metrics are actually coming from reflectors of CRD resources. Around 15000 of those metrics are each related to one of the three CRDs installed in the cluster. The total number of metrics increases by ~50 every few minutes.

I feel this might be happening whenever the cache in the Prometheus Operator (the controller for the CRDs existing in this cluster) is hitting its resync interval for a particular resource, so this might be happening on either opening a watch and or listing the resources, but this is just a wild guess.

The metrics in the `NewNamedReflector` function should be injected instead of re-created and registered on the global metrics registry, but the fact that the apiserver keeps creating new reflectors for these resources is likely also a problem.

Another problem I've been noticing, starting approximately at the same time as this, is that the apiserver API latencies have increased significantly.

**What you expected to happen**:

Unique metrics for each type of reflector, not causing a memory leak.

**How to reproduce it (as minimally and precisely as possible)**:

The cluster was setup using bootkube, but recently upgraded to 1.8 (3 days ago), and the Prometheus Operator running on it with the [`kube-prometheus`](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus) stack deployed.

**Anything else we need to know?**:

**Environment**: bootkube / bare metal
- Kubernetes version (use `kubectl version`): 1.8
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release): CoreOS Container Linux
- Kernel (e.g. `uname -a`): Linux sd-99796 4.12.14-coreos #1 SMP Wed Sep 20 22:20:05 UTC 2017 x86_64
- Install tools: bootkube
- Others:

@sig-instrumentation-bugs @kubernetes/sig-api-machinery-bugs 

/cc @sttts @nikhita I know you recently worked on the CRD code, so you may know why there are new reflectors being created constantly.

### PR DESCRIPTION
**Pull request by sttts**
Potentially fixes #53485

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -426,10 +427,13 @@ func (c *crdHandler) updateCustomResourceDefinition(oldObj, _ interface{}) {
...
 	// as it is used without locking elsewhere
 	for k, v := range storageMap {
+		if k == oldCRD.UID {
+			v.storage.DestroyFunc()
+			continue
+		}
 		storageMap2[k] = v
 	}
 
-	delete(storageMap2, oldCRD.UID)
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: why isn't this delete needed anymore?

**Comment by deads2k** [comment]
question, then lgtm.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Because of the `continue` above.

**Comment by sttts** [comment]
@deads2k @brancz added a commit which more conservatively updates the storage. I.e. only when spec or status.acceptedNames change, we update anything. @brancz found out that we leak prometheus metrics with every new storage. This PR does not fix the actual leak (https://github.com/kubernetes/kubernetes/issues/52121), but it considerably reduces its effect if for some reason CRDs are changed very often (e.g. via annotations or other status updates).

**Comment by sttts** [comment]
There is a race in CRD deletion that this PR made visible. I have a fix for it in the PR context, but would like to fix the race in general.

**Comment by sttts** [comment]
@ironcladlou @deads2k @caesarxuchao the shutdown of the storage triggers the following "issue" with the GC controller: when the storage is shutdown, the watches die and the GC controller has to relist the objects. This is fine during normal operation, i.e. when a CRD is just updated, but is kept alive. It is not so clear when we are terminating the CRD:

Comment out https://github.com/kubernetes/kubernetes/pull/53586/files#diff-2f720a7e61a4ddfcfe9a9e353be8784bR441. Then:

1. we delete the CRD by setting the deletionTimestamp
2. the CR storage is shutdown because the CR handler sees the CRD update (this PR). All CR watches are killed by the storage shutdown.
3. the apiextension server will delete all corresponding CRs and finally deletes the CRD. At that point the CR api endpoint is gone, giving 404.
4. the GC controller might not have seen yet the deletions of CRs, but it noticed the CRD update and rediscovers resources.
5. the GC controller notices the CRD is gone and it will remove its watch (which possibly tried to relist unsuccessfully or is already dead because the CR storage in the apiserver was shutdown).
6. **I suspect** the GC controller will not mark the corresponding CRs as gone. Hence, the ownees of the CRs will not be garbage collected.

If we keep the return at https://github.com/kubernetes/kubernetes/pull/53586/files#diff-2f720a7e61a4ddfcfe9a9e353be8784bR441, step 2 will not take place until the CRD is deleted from storage (step 3). At that time it seems that the watches on the CRs are already notified about the CR deletions, i.e. the GC controller has seen the CRs go and the ownees are correctly garabage collected.

Though, I am not 100% sure that we don't have a race here and we can still see the GC missing the deletion of the CRs.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -410,23 +424,36 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
...
 	oldCRD := oldObj.(*apiextensions.CustomResourceDefinition)
-	glog.V(4).Infof("Updating customresourcedefinition %s", oldCRD.Name)
+	newCRD := oldObj.(*apiextensions.CustomResourceDefinition)
 
 	c.customStorageLock.Lock()
 	defer c.customStorageLock.Unlock()
-
 	storageMap := c.customStorage.Load().(crdStorageMap)
+
+	oldInfo, found := storageMap[newCRD.UID]
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Shouldn't this be `oldCRD.UID`?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: the UID does not change, i.e. it does not matter here.

**Comment by caesarxuchao** [comment]
> I suspect the GC controller will not mark the corresponding CRs as gone

IIRC, it won't. It's to make GC safe for user apiserver. If an UAS get network partitioned and is temporarily removed from the discovery API, we don't want to remove the ownees prematurely. cc @ironcladlou

**Comment by sttts** [comment]
@nikhita ptal at the test fix

**Comment by nikhita** [comment]
@sttts works great now! :tada:

**Comment by brancz** [comment]
Thanks for taking care of this @sttts! I'll try to deploy this on my cluster as well to validate, that it fixes the reported issue.

Can we make sure, that this is cherry-picked to land in the next minor release of 1.8?

**Comment by sttts** [comment]
@brancz yes, please try in your cluster. If it helps, I think we should cherry-pick it to 1.8.

**Pull request merged**